Introducing touch events and scrolling containers touching their content.

reven86 opened the pull request:
This adds support for processing touch events, especially on mobile devices. Containers with overflow set to auto or scroll can now be scrolled using touches, also applies inertia when using swipe gesture - releasing the finger while scrolling.

What is done:
- added ProcessTouch* events as suggested in https://github.com/mikke89/RmlUi/issues/60
- added new scrolling mode for ScrollController
- handles up to 10 touches simultaneously

What is not done:
- no touchstart, touchmove, etc. events propagated to Elements
- while you can scroll two or more containers simultaneously moving your fingers, inertia is applied to only one of them since ScrollController handles scrolling for only one element at a time
- no samples and backends were modified to submit ProcessTouch* events to Rml

wh1t3lord commented on Include/RmlUi/Core/Context.h:
```diff
@@ -344,6 +363,22 @@ class RMLUICORE_API Context : public ScriptInterface {
+		double scrolling_last_time = 0;
+		double scrolling_start_time_x = 0;
+		double scrolling_start_time_y = 0;
+		bool scrolling_right = false;
```
@reven86  Hi and thanks for PR, but I suggest to move these bools to first rows and improve memory layout of struct for faster construction. It might seem as a pointless but for real the time goes and code base consists of tiny details and if all details would be written as not good then whole application would be not good and laggy. Thanks 

<img width="372" height="141" alt="image" src="https://github.com/user-attachments/assets/04d3b793-32cf-4f0c-888c-ccea9f4e3f41" />

reven86 replied:
I don't see any performance gain here, but OK. The memory layout indeed is few bytes smaller.

mikke89 commented:
It's great to see this being worked, thanks a lot! It's very nice to see that you use the scroll controller for inertial scrolling. 

I think it's perfectly fine to not implement the actual RML events, especially for a first iteration. We should ensure though that the new API on the context is compatible with events if we want to add that in the future. Generally, we try to adhere to HTML events, unless there are good reasons not to. If we look a the [Web Touch events](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events), we see that single events are emitted with a group of touches. Looks like this wouldn't really work that well when submitting one and one touch point at a time to the context, like here.

My suggestion would instead be to take a list of touches on each of the new context functions. What do you think?

I think it would also be preferable to allow subpixel positioning, e..g using a Vector2f for that. And I think it would be nice to allow the user to provide an arbitrary integer ID, without fixing it to 10.

mikke89 commented:
> How about making a cmake option, something like RMLUI_BACKEND_MOUSE_AS_TOUCHES

Yeah, I think that's a great idea. That would help with testing and reviewing things too. My suggestion would be `RMLUI_BACKEND_SIMULATE_TOUCH`.

reven86 commented:
@mikke89  I've implemented and tested `RMLUI_BACKEND_SIMULATE_TOUCH` with sample projects on Win32 and GLFW platforms. I can't test it on other platforms so it's left unimplemented there.

mikke89 commented on Source/Core/ScrollController.cpp:
```diff
@@ -190,6 +221,32 @@ void ScrollController::UpdateSmoothscroll(float dp_ratio)
+	Vector2f scroll_delta = inertia_scroll_velocity * dt;
+	PerformScrollOnTarget(scroll_delta);
+
+	float dampening = 1.0f - INERTIA_FRICTION_FACTOR * dt;
```
We should handle small and large dt here, to ensure stability, e.g. by clamping the dampening to some constants larger than zero and smaller than one.

reven86 replied:
Sure, it is clamped already during `ScrollController::UpdateTime`

mikke89 commented on Source/Core/ScrollController.cpp:
```diff
@@ -123,12 +126,40 @@ void ScrollController::ActivateSmoothscroll(Element* in_target, Vector2f delta_d
+	if (!in_target)
+		return;
+
+	// instant scroll element without changing the current target
```
Is this really necessary - should perhaps an instant scroll cancel any existing autoscroll?

reven86 replied:
Yes, since ScrollController can handle only one smooth/auto/inertia scrolling at the moment I don't want to break it when we need an instant scroll on another container. Ideally ScrollController should support scrolling multiple elements at once, but this probably is better to make as another PR.

mikke89 commented on CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ option(BUILD_SHARED_LIBS "CMake standard option. Choose whether to build shared
 #   - Do not include negations (such as "not" and "disable"), to avoid situations with double negation.
 #   - Do not include a verb prefix (such as "enable" and "build"), as these are often superfluous.
 option(RMLUI_SAMPLES "Build samples of the library." OFF)
+option(RMLUI_BACKEND_SIMULATE_TOUCH "Simulate touch events with mouse events" OFF)
```
Great to see that you added this option. I think we should error out though if the backend does not support touch simulation and this option is set.

mikke89 commented on Source/Core/Context.cpp:
```diff
@@ -875,6 +878,175 @@ bool Context::IsMouseInteracting() const
+	state->scroll_container = nullptr;
+
+	ProcessMouseMove(static_cast<int>(touch.position.x), static_cast<int>(touch.position.y), 0);
+	return ProcessMouseButtonUp(static_cast<int>(std::distance(&touch_states.front(), state)), 0);
```
One quirk I found during simulation is that the mouse cursor is only set after clicking an element, and then it stays persistent. Maybe this is just an acceptable quirk of simulation. Ideally we would disable the mouse while touching, as if calling `ProcessMouseLeave`, and also the cursor. But I guess this wouldn't really work directly now that we use mouse events under the hood.

I would say this one is not a blocker.

mikke89 reviewed (commented):
Thanks a lot! Very nice work. I've reviewed everything now, and tested it more thoroughly on a touch-screen display. And I find that the scrolling feels really nice. It's nice that 2D-scrolling works fluently as well.

I found one behavioral part that could perhaps be improved. When quickly dragging, then stopping for a small moment, and letting go. Then the scrolling speed will suddenly jump when I expected zero autoscrolling. The reverse also applies similarly: If you move very slowly in one direction for a while, then quickly swipe up and release, then velocity will be lower than expected. It's not a blocker, we can improve it in follow-ups.

Another potential follow-up is to make it so that we don't apply hover on elements during touch movement. That feels a bit weird, and I see that browser don't apply it as well. Ref. comment about `ProcessMouseLeave`.

---

As for the backend changes. Maybe it's tricky to avoid, but I'm not super-thrilled about all the intricate ifdefs. I wonder if we can make it a bit cleaner by collecting the ifdef parts into reusable components, gathered into one or a few ifdefs, where practical?

Do you think it would be a lot of effort to implement touch support on some of these backends? Perhaps we could reuse some of the touch simulation logic, and avoid some of the ifdefs then. I have a (multi-) touch screen display on a laptop, so it would be cool to test it there. Some of the backends should work on Android and other mobile platforms (possibly with smaller changes), so users might take advantage of this support directly. And there are plenty of desktop/laptop/tablet environments with touch screens too.

Unfortunately I see that GLFW [doesn't yet support touch input](https://github.com/glfw/glfw/issues/42). Though I see that SDL does have native support for touch, which seems really nice. And even a [built-in way to simulate them](https://wiki.libsdl.org/SDL3/SDL_HINT_MOUSE_TOUCH_EVENTS). Maybe that's the way to go, to implement the feature properly, and simply turn on its built-in simulation where possible?

It's okay if you want to skip these implementations to get the PR in for the core features. If so, I would suggest to keep the simulation logic in just a single platform, as a temporary testing solution until we can add proper touch support.

reven86 commented:
> I found one behavioral part that could perhaps be improved. When quickly dragging, then stopping for a small moment, and letting go. Then the scrolling speed will suddenly jump when I expected zero autoscrolling. The reverse also applies similarly: If you move very slowly in one direction for a while, then quickly swipe up and release, then velocity will be lower than expected. It's not a blocker, we can improve it in follow-ups.

I see, that's because we use average velocity for the entire time period between touch start and touch end events, while it is better to use some small time frame to calculate velocity. I think applying some weight to the start position proportionally to the elapsed time should do the trick. I'll try that.

> Another potential follow-up is to make it so that we don't apply hover on elements during touch movement. That feels a bit weird, and I see that browser don't apply it as well. Ref. comment about `ProcessMouseLeave`.

That's a little harder to fix as all our touch events are internally processed as mouse events and I guess it will require checking the exact source of the event (was it a mouse or a touch?) when updating hover element.

> As for the backend changes. Maybe it's tricky to avoid, but I'm not super-thrilled about all the intricate ifdefs. I wonder if we can make it a bit cleaner by collecting the ifdef parts into reusable components, gathered into one or a few ifdefs, where practical?

I don't like it either, I'll try to rearrange and group them.

> Do you think it would be a lot of effort to implement touch support on some of these backends? Perhaps we could reuse some of the touch simulation logic, and avoid some of the ifdefs then. I have a (multi-) touch screen display on a laptop, so it would be cool to test it there. Some of the backends should work on Android and other mobile platforms (possibly with smaller changes), so users might take advantage of this support directly. And there are plenty of desktop/laptop/tablet environments with touch screens too.

I currently can only test the app on GLFW and Win32 backends. GLFW doesn't support touches natively and for Win32 I don't have device with touch support. I can implement handling WM_TOUCH message on Win32 but can't properly test it.

reven86 commented:
@mikke89 I've addressed your remarks except of ones related to touch simulation and I will need your opinion here. At first I had an idea to implement touch simulation logic mostly for debugging purposes, because in fact you only need a simulation when developing for touch devices on desktop to safe time deploying and testing on actual touch devices. For that reason I thought the right place to implement the simulation logic would be in backends.

However, implementing it there turned out in using not so clean solutions which is hard to avoid: ifdefs when handling mouse events, global state which should be exposed somehow to platform callbacks either by being part of Backend's public interface or using global variables, some code repetition used in different backends.

It all can be solved by introducing a thin wrapper over Context that intercepts mouse events and transforms them internally into touch events, or implementing it directly in the Context, and this will also work for all backends with no changes to a platform code. One thing I also realized is that the state required for simulation (e.g. `touch_simulation_state` in my code) should not be global, but should also be tied to a particular context, since we can have multiple contexts in the app and each of them should process mouse and touch events independently.

So the best option here IMO is to make touch simulation logic right inside the Context object, but I'm not very happy to put it there bearing in mind it is only needed as development/debugging option.

mikke89 commented on Source/Core/Context.cpp:
```diff
@@ -875,6 +878,175 @@ bool Context::IsMouseInteracting() const
+		state->identifier = touch.identifier;
+
+		RMLUI_ASSERTMSG(touch_states.size() < 100, "Too many different touch identifiers were used.");
+	}
```
I don't think this is resolved fully. In particular, I don't see the touch being removed. I am a bit concerned for cases where a new touch id is assigned for every new touch occurrence, thus growing the map infinitely.

mikke89 commented:
Excellent stuff, looks solid to me! This is a very welcome addition to the library, so thanks a lot.

I will follow-up with the SDL simulation changes. I'll also start updating the documentation with these changes.
